File: frontend/src/app/services/customer-device.service.ts
import { Injectable } from '@angular/core';
import { HttpClient, HttpParams } from '@angular/common/http';
import { Observable } from 'rxjs';
import { CustomerDevice } from '../models/customer-device.model';
import { Page } from '../models/page.model';

@Injectable({
  providedIn: 'root'
})
export class CustomerDeviceService {
  private apiUrl = 'http://localhost:8080/api/customer-devices';

  constructor(private http: HttpClient) { }

  getAllCustomerDevices(page: number = 0, size: number = 25): Observable<Page<CustomerDevice>> {
    const params = new HttpParams()
      .set('page', page.toString())
      .set('size', size.toString());
    return this.http.get<Page<CustomerDevice>>(this.apiUrl, { params });
  }

  getCustomerDeviceById(id: number): Observable<CustomerDevice> {
    return this.http.get<CustomerDevice>(`${this.apiUrl}/${id}`);
  }

  getCustomerDevicesByCustomerId(customerId: string): Observable<CustomerDevice[]> {
    return this.http.get<CustomerDevice[]>(`${this.apiUrl}/customer/${customerId}`);
  }

  getCustomerDevicesBySerial(serial: string): Observable<CustomerDevice[]> {
    return this.http.get<CustomerDevice[]>(`${this.apiUrl}/serial/${serial}`);
  }

  getCustomerDevicesByStatus(status: string): Observable<CustomerDevice[]> {
    return this.http.get<CustomerDevice[]>(`${this.apiUrl}/status/${status}`);
  }

  createCustomerDevice(customerDevice: CustomerDevice): Observable<CustomerDevice> {
    return this.http.post<CustomerDevice>(this.apiUrl, customerDevice);
  }

  updateCustomerDevice(id: number, customerDevice: CustomerDevice): Observable<CustomerDevice> {
    return this.http.put<CustomerDevice>(`${this.apiUrl}/${id}`, customerDevice);
  }

  deleteCustomerDevice(id: number): Observable<void> {
    return this.http.delete<void>(`${this.apiUrl}/${id}`);
  }
}

File: backend/src/main/java/com/teamcommercial/service/CustomerDeviceService.java
package com.teamcommercial.service;

import com.teamcommercial.entity.CustomerDevice;
import com.teamcommercial.repository.CustomerDeviceRepository;
import org.springframework.beans.factory.annotation.Autowired;
import org.springframework.data.domain.Page;
import org.springframework.data.domain.Pageable;
import org.springframework.stereotype.Service;
import org.springframework.transaction.annotation.Transactional;

import java.util.List;
import java.util.Optional;

@Service
@Transactional
public class CustomerDeviceService {

    private final CustomerDeviceRepository customerDeviceRepository;

    @Autowired
    public CustomerDeviceService(CustomerDeviceRepository customerDeviceRepository) {
        this.customerDeviceRepository = customerDeviceRepository;
    }

    public Page<CustomerDevice> getAllCustomerDevices(Pageable pageable) {
        return customerDeviceRepository.findAll(pageable);
    }
    
    public List<CustomerDevice> getAllCustomerDevices() {
        return customerDeviceRepository.findAll();
    }

    public Optional<CustomerDevice> getCustomerDeviceById(Long id) {
        return customerDeviceRepository.findById(id);
    }

    public List<CustomerDevice> getCustomerDevicesByCustomerId(String customerId) {
        return customerDeviceRepository.findByCustomerId(customerId);
    }

    public List<CustomerDevice> getCustomerDevicesBySerial(String serial) {
        return customerDeviceRepository.findBySerial(serial);
    }

    public List<CustomerDevice> getCustomerDevicesByStatus(String status) {
        return customerDeviceRepository.findByStatus(status);
    }

    public CustomerDevice createCustomerDevice(CustomerDevice customerDevice) {
        return customerDeviceRepository.save(customerDevice);
    }

    public CustomerDevice updateCustomerDevice(Long id, CustomerDevice customerDeviceDetails) {
        CustomerDevice customerDevice = customerDeviceRepository.findById(id)
                .orElseThrow(() -> new RuntimeException("Customer Device not found with id: " + id));
        
        customerDevice.setCustomerId(customerDeviceDetails.getCustomerId());
        customerDevice.setSerial(customerDeviceDetails.getSerial());
        customerDevice.setControllerDevice(customerDeviceDetails.getControllerDevice());
        customerDevice.setStatus(customerDeviceDetails.getStatus());
        
        return customerDeviceRepository.save(customerDevice);
    }

    public void deleteCustomerDevice(Long id) {
        CustomerDevice customerDevice = customerDeviceRepository.findById(id)
                .orElseThrow(() -> new RuntimeException("Customer Device not found with id: " + id));
        customerDeviceRepository.delete(customerDevice);
    }
}

